pretrain/tinyllama.py

```diff
@@ -0,0 +1,340 @@
+from lightning.fabric.loggers import CSVLogger
+from lightning.pytorch.loggers import WandbLogger
+from torch.utils.data import DataLoader
+
+# support running without installing as a package
+wd = Path(__file__).parent.parent.resolve()
+sys.path.append(str(wd))
+
+from lit_gpt.model import GPT, Block, Config, LLaMAMLP
+from lit_gpt.packed_dataset import CombinedDataset
+from lit_gpt.utils import chunked_cross_entropy, num_parameters
+
+# System settings
+model_name = "tiny-llama-1.1b"
+name = "lit-tiny-llama-1.1b"
+out_dir = Path("out") / name
+use_wandb = False
+
+# Hyperparameters
+devices = 4  # TODO: set this to 8
+
+global_batch_size = 32  # TODO: should be 512
+learning_rate = 4e-4
+micro_batch_size = 1  # TODO: should be 8
+max_step = 715256 * 2
+warmup_steps = 2000
+log_step_interval = 2
+eval_iters = 100
+save_step_interval = 1000
+eval_step_interval = 1000
+
+weight_decay = 1e-1
+beta1 = 0.9
+beta2 = 0.95
+grad_clip = 1.0
+decay_lr = True
+min_lr = 4e-5
+
+batch_size = global_batch_size // devices
+gradient_accumulation_steps = batch_size // micro_batch_size
+assert gradient_accumulation_steps > 0
+warmup_iters = warmup_steps * gradient_accumulation_steps
+
+max_iters = max_step * gradient_accumulation_steps
+lr_decay_iters = max_iters
+log_iter_interval = log_step_interval * gradient_accumulation_steps
+
+
+hparams = {k: v for k, v in locals().items() if isinstance(v, (int, float, str)) and not k.startswith("_")}
+
+
+def setup(resume: Union[bool, Path] = False):
+    if use_wandb:
+        logger = WandbLogger(project="tinyllama", name="training", resume=(resume is not False))
+    else:
+        logger = CSVLogger(root_dir="logs", name="tinyllama")
+
+    if devices > 1:
+        strategy = FSDPStrategy(
+            auto_wrap_policy={Block},
+            activation_checkpointing_policy=None,
+            state_dict_type="full",
+            limit_all_gathers=True,
+            cpu_offload=False,
+        )
+    else:
+        strategy = "auto"
+
+    fabric = L.Fabric(devices=devices, strategy=strategy, precision="bf16-mixed", loggers=[logger])
+    fabric.launch()
+
+    fabric.print(hparams)
+    if use_wandb:
+        logger.log_hyperparams(hparams)
+
+    main(fabric, resume)
+
+
+def main(fabric, resume):
+    if fabric.global_rank == 0:
+        out_dir.mkdir(parents=True, exist_ok=True)
+
+    config = Config.from_name(model_name)
+
+    train_dataloader, val_dataloader = create_dataloaders(
+        fabric, batch_size=micro_batch_size, block_size=config.block_size
+    )
+    if val_dataloader is None:
+        train_dataloader = fabric.setup_dataloaders(train_dataloader)
+    else:
+        train_dataloader, val_dataloader = fabric.setup_dataloaders(train_dataloader, val_dataloader)
+
+    fabric.seed_everything(3407)  # same seed for every process to init model (FSDP)
+
+    fabric.print(f"Loading model with {config.__dict__}")
+    t0 = time.perf_counter()
+    with fabric.init_module(empty_init=False):
```

Comment:
In some scripts `empty_init` depends on the number of devices. The same is true for the TinyLLaMA repo. 
Not sure why there is an inconsistency.

Reply:
I need to init the weights with particular distributions. This isn't possible if the model is on the meta device. The model is 1.1b, so it won't be a problem to initialize. It will be a bit slower.

Reply:
@awaelchli I think @Andrei-Aksionov is right there is a fix for `empty_init` [here](https://github.com/jzhang38/TinyLlama/pull/89/files)

Reply:
No, the way it is in tinyllama is incorrect. The line

```
 model.apply(partial(model._init_weights ,n_layer=config.n_layer))
 ```
 
 won't have any affect when the model is on the meta device!

Reply:
Got it!
Thanks

Reply:
@carmocca Thank you so much! I will have a carefully look at this